faiss_vector_cache.go

```diff
@@ -0,0 +1,206 @@
+//  Copyright (c) 2024 Couchbase, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 		http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//go:build vectors
+// +build vectors
+
+package zap
+
+import (
+	"sync"
+	"sync/atomic"
+	"time"
+
+	faiss "github.com/blevesearch/go-faiss"
+)
+
+type vecCache struct {
+	closeCh chan struct{}
+	m       sync.RWMutex
+	cache   map[uint16]*cacheEntry
+}
+
+type ewma struct {
+	alpha  float64
+	avg    float64
+	sample uint64
+}
+
+type cacheEntry struct {
+	tracker *ewma
+
+	m     sync.RWMutex
+	index *faiss.IndexImpl
+}
+
+func newVectorCache() *vecCache {
+	return &vecCache{
+		cache:   make(map[uint16]*cacheEntry),
+		closeCh: make(chan struct{}),
+	}
+}
+
+func (vc *vecCache) Clear() {
+	vc.m.Lock()
+	close(vc.closeCh)
+	vc.cache = nil
+	vc.m.Unlock()
+
+}
+
+func (vc *vecCache) checkCacheForVecIndex(fieldID uint16,
+	indexBytes []byte) (vecIndex *faiss.IndexImpl, err error) {
+	cachedIndex, present := vc.isVecIndexCached(fieldID)
+	if present {
+		vecIndex = cachedIndex
+		vc.addRef(fieldID)
+	} else {
+		// if the cache doesn't have vector index, just construct it out of the
+		// index bytes and update the cache.
+		vecIndex, err = faiss.ReadIndexFromBuffer(indexBytes, faiss.IOFlagReadOnly)
+		vc.update(fieldID, vecIndex)
+	}
+	return vecIndex, err
+}
+
+func (vc *vecCache) isVecIndexCached(fieldID uint16) (*faiss.IndexImpl, bool) {
+	vc.m.RLock()
+	entry, present := vc.cache[fieldID]
+	vc.m.RUnlock()
+	if entry == nil {
+		return nil, false
+	}
+
+	entry.m.RLock()
+	rv := entry.index
+	entry.m.RUnlock()
+
+	return rv, present && (rv != nil)
+}
+
+func (vc *vecCache) addRef(fieldIDPlus1 uint16) {
+	vc.m.RLock()
+	entry := vc.cache[fieldIDPlus1]
+	vc.m.RUnlock()
+
+	entry.addRef()
+}
+
+func (vc *vecCache) clearField(fieldIDPlus1 uint16) {
+	vc.m.Lock()
+	delete(vc.cache, fieldIDPlus1)
+	vc.m.Unlock()
+}
+
+func (vc *vecCache) refreshEntries() (rv int) {
+	vc.m.RLock()
+	cache := vc.cache
```

Comment:
This is racey isn't it - you're grabbing a reference so a data race could occur in updating this map.

Reply:
done